storage/HashTableFactory.hpp

```diff
@@ -305,10 +309,37 @@ class HashTableFactory {
       key_types.emplace_back(&TypeFactory::ReconstructFromProto(proto.key_types(i)));
     }
 
-    return CreateResizable(HashTableImplTypeFromProto(proto.hash_table_impl_type()),
-                           key_types,
-                           proto.estimated_num_entries(),
-                           storage_manager);
+    auto hash_table = CreateResizable(HashTableImplTypeFromProto(proto.hash_table_impl_type()),
+                                      key_types,
+                                      proto.estimated_num_entries(),
+                                      storage_manager);
+
+    // Check if there are any build side bloom filter defined on the hash table.
+    if (proto.build_side_bloom_filter_id_size() > 0) {
+      hash_table->enableBuildSideBloomFilter();
+      hash_table->setBuildSideBloomFilter(bloom_filters[proto.build_side_bloom_filter_id(0)].get());
+    }
+
+    // Check if there are any probe side bloom filters defined on the hash table.
+    if (proto.probe_side_bloom_filters_size() > 0) {
+      hash_table->enableProbeSideBloomFilter();
+      // Add as many probe bloom filters as defined by the proto.
+      for (int j = 0; j < proto.probe_side_bloom_filters_size(); ++j) {
+        // Add the pointer to the probe bloom filter within the list of probe bloom filters to use.
+        const auto probe_side_bloom_filter = proto.probe_side_bloom_filters(j);
+        hash_table->addProbeSideBloomFilter(bloom_filters[probe_side_bloom_filter.probe_side_bloom_filter_id()].get());
+
+        // Add the attribute ids corresponding to this probe bloom filter.
+        std::vector<attribute_id> probe_attribute_ids;
+        for (int k = 0; k < probe_side_bloom_filter.probe_side_attr_ids_size(); ++k) {
+          const attribute_id probe_attribute_id = probe_side_bloom_filter.probe_side_attr_ids(k);
+          probe_attribute_ids.push_back(probe_attribute_id);
+        }
+        hash_table->addProbeSideAttributeIds(std::move(probe_attribute_ids));
+      }
+    }
```

Comment:
What do you think of moving all these late-initializations into `CreateResizable` and the hash table constructors? With this being said, we don't need additional helper methods in `HashTable.hpp`, and could mark these data members `const` wherever possible.

```
    bool has_build_side_bloom_filter = false;
    BloomFilter *build_bloom_filter = nullptr;
    if (proto.build_side_bloom_filters_size() > 0) {
      has_build_side_bloom_filter = true;
      build_bloom_filter = bloom_filters[proto.build_side_bloom_filter_id(0)].get();
    }

    const bool has_probe_side_bloom_filter = proto.probe_side_bloom_filters_size() > 0;
    std::vector<const BloomFilter*> probe_bloom_filters(proto.probe_side_bloom_filters_size());
    std::vector<std::vector<attribute_id>> probe_attribute_ids_list;
    for (int i = 0; i < proto.probe_side_bloom_filters_size(); ++i) {
      const auto probe_side_bloom_filter = proto.probe_side_bloom_filters(i);
      probe_bloom_filters[i] = bloom_filters[probe_side_bloom_filter.probe_side_bloom_filter_id()].get();

      std::vector<attribute_id> probe_attribute_ids(probe_side_bloom_filter.probe_side_attr_ids_size());
      for (int j = 0; j < probe_side_bloom_filter.probe_side_attr_ids_size(); ++j) {
        probe_attribute_ids[j] = probe_side_bloom_filter.probe_side_attr_ids(j);
      }
      probe_attribute_ids_list.push_back(std::move(probe_attribute_ids));
    }

    return CreateResizable(HashTableImplTypeFromProto(proto.hash_table_impl_type()),
                           key_types,
                           proto.estimated_num_entries(),
                           has_build_side_bloom_filter,
                           build_bloom_filter,
                           has_probe_side_bloom_filter,
                           std::move(probe_bloom_filters),
                           std::move(probe_attribute_ids_list),
                           storage_manager);
```

We also need to modify various hash table constructors, i.e., HashTable below:

```
// storage/HashTable.hpp
  HashTable(...,
            const bool has_build_side_bloom_filter,
            BloomFilter *build_bloom_filter,
            const bool has_probe_side_bloom_filter,
            std::vector<const BloomFilter*> &&probe_bloom_filters,
            std::vector<std::vector<attribute_id>> &&probe_attribute_ids)
      : ...
        has_build_side_bloom_filter_(has_build_side_bloom_filter),
        build_bloom_filter_(build_bloom_filter),
        has_probe_side_bloom_filter_(has_probe_side_bloom_filter),
        probe_bloom_filters_(std::move(probe_bloom_filters)),
        probe_attribute_ids_(std::move(probe_attribute_ids)) {}

...

  const bool has_build_side_bloom_filter_;
  BloomFilter *build_bloom_filter_;

  const bool has_probe_side_bloom_filter_;
  const std::vector<const BloomFilter*> probe_bloom_filters_;
  const std::vector<std::vector<attribute_id>> probe_attribute_ids_;
```